File: 2nd_sem/code_for_presentation/dip/GoodPractice.kt
interface Database {
    fun saveData(data: String)
}

class MySQLDatabase : Database {
    override fun saveData(data: String) {
        println("Зберігаю в MySQL: $data")
    }
}

class PostgreSQLDatabase : Database {
    override fun saveData(data: String) {
        println("Зберігаю в PostgreSQL: $data")
    }
}

class FileStorage : Database {
    override fun saveData(data: String) {
        File("report.txt").writeText(data)
        println("Зберігаю в файл: $data")
    }
}

class ReportService(private val database: Database) { // Залежність через інтерфейс!
    fun generateReport() {
        database.saveData("Звіт 2024")
    }
}



// dependency injection
fun main() {
    val mySqlReport = ReportService(MySQLDatabase())
    mySqlReport.generateReport() // "Зберігаю в MySQL: Звіт 2024"

    val postgresReport = ReportService(PostgreSQLDatabase())
    postgresReport.generateReport() // "Зберігаю в PostgreSQL: Звіт 2024"

    val fileReport = ReportService(FileStorage())
    fileReport.generateReport() // Зберігає у файл "report.txt"
}

File: 2nd_sem/Presentations/soLID/dip/GoodPractice.cs
// easy-expandable code
public interface IDatabase
{
    void SaveData(string data);
}

public class MySQLDatabase : IDatabase
{
    public void SaveData(string data)
    {
        Console.WriteLine($"Зберігаю в MySQL: {data}");
    }
}

public class PostgreSQLDatabase : IDatabase
{
    public void SaveData(string data)
    {
        Console.WriteLine($"Зберігаю в PostgreSQL: {data}");
    }
}

public class FileStorage : IDatabase
{
    public void SaveData(string data)
    {
        File.WriteAllText("report.txt", data);
        Console.WriteLine($"Зберігаю в файл: {data}");
    }
}

public class ReportService
{
    private readonly IDatabase _database;

    public ReportService(IDatabase database)
    { // Залежність через інтерфейс!
        _database = database;
    }

    public void GenerateReport()
    {
        _database.SaveData("Звіт 2024");
    }
}

// dependency injection
var mySqlReport = new ReportService(new MySQLDatabase());
mySqlReport.GenerateReport(); // "Зберігаю в MySQL: Звіт 2024"

var postgresReport = new ReportService(new PostgreSQLDatabase());
postgresReport.GenerateReport(); // "Зберігаю в PostgreSQL: Звіт 2024"

var fileReport = new ReportService(new FileStorage());
fileReport.GenerateReport(); // Зберігає у файл "report.txt"
